### DISCUSSION TITLE
Questions about embedding deno

### DISCUSSION THREAD

**Discussion by augustocdias** [denoland/deno#9595] [Q&A]
Hello,

I have some questions about embedding deno that I couldn't find in any docs:

1. Which crate should I use: `deno_core` or `deno_runtime`? 
2. Why the `execute` method in both runtime and core is not a future? I suppose the `run_event_loop` is for when I want to run something like a server written in javascript that will run indefinitely, right?
3. Is it somehow possible to have imports like in the deno manual, pointing to urls instead of local files? When I tried with the runtime it raised an error saying it wasn't a file. 
4. I suppose the typescript support is not available neither in the runtime nor in the core right?

**Comment by bartlomieju** [comment]
> Hello,
> 
> I have some questions about embedding deno that I couldn't find in any docs:
> 
> 1. Which crate should I use: `deno_core` or `deno_runtime`?

It depends if you just need basic building blocks, or want to use APIs provided in `Deno` namespace; `deno_runtime` has implementation of most `Deno` APIs (sans runtime TS compiler).

> 2. Why the `execute` method in both runtime and core is not a future? I suppose the `run_event_loop` is for when I want to run something like a server written in javascript that will run indefinitely, right?

`JsRuntime::execute` is used for executing plain scripts (ie. not ES modules), execution of plain scripts is synchronous. You might want to look into `JsRuntime::load_module` and `JsRuntime::mod_evaluate` which are async and used for ES modules loading. 

`JsRuntime::run_event_loop` is a future that resolves when there are no more async ops pending and users are expected to call it manually after executing script or module.

> 3. Is it somehow possible to have imports like in the deno manual, pointing to urls instead of local files? When I tried with the runtime it raised an error saying it wasn't a file.

URL import support is not part of `deno_runtime` crate, it's implemented in the `cli` directory and is currently not available to embedders.

> 4. I suppose the typescript support is not available neither in the runtime nor in the core right?

Correct, same story as with URL import support.

For more info, please drop by to Deno's Discord server (https://discord.gg/deno), issues are meant to be used for bug reports.

**Reply by twlite** [reply]
hi, what would be the status of this today?

**Reply by bartlomieju** [reply]
What are you asking about specifically?

**Reply by twlite** [reply]
I want to embed deno in my application. I went to deno manual but it lacks explanation about these details. After some digging, I ended up here. I was wondering what things have changed since 2021 and which of your points are same/which points have changed, etc.

**Reply by bartlomieju** [reply]
Points 3 and 4 have not changed. They are still part of the `deno` crate that is not available as a library and if you wish to use them you will have to implement the system yourself.
